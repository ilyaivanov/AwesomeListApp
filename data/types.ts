
export type Section = {
  title: string;
  id: string;
  links: Link[];
}

export type Link = {
  title: string;
  link: string;
  subtitle?: string,
  level?:number;
}

export {Token} from 'markdown-it';