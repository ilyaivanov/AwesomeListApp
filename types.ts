export {Token} from 'markdown-it';

export interface Section {
  title: string;
  links: Link[];
}

export interface Link {
  title: string;
  subtitle: string,
  link: string,
  level:number;
}

export interface Repository {
  home: Section,
  sections: Section[],
}