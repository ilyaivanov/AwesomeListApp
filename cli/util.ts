import {flatten} from 'lodash';
import {Repository} from '../types';

export const getParts = (url: string) => {
  const parts = url.split('/');
  return parts.slice(parts.length - 2);
}

export const toMd = (url: string) => {
  const [username, reponame] = getParts(url);
  return `https://api.github.com/repos/${username}/${reponame}/readme`;
};

export const createFilePath = (base: string, url: string, extension: string) =>
  base + getParts(url).join('_') + '.' + extension;


export const allLinks = (rep: Repository) =>
  flatten(rep.sections.map(s => s.links.map(l => l.link)))


export const decode = (response: string, encoding: string) => new Buffer(response, encoding).toString('utf8');

export const stringify = (obj: any) => JSON.stringify(obj, null, 2);

export const root = {
  name: 'Awesome',
  url: 'https://github.com/sindresorhus/awesome'
};

