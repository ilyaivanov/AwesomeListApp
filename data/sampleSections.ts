const sections: Section[] = [
  {
    title: 'Awesome List (dummy)',
    id: 'root',
    links: [
      {title: 'Platforms', link: '#platforms'},
      {title: 'Languages', link: '#languages', subtitle:'fancy languages'},
      {title: 'Functional Languages', link: '#functional', subtitle:'subsection', level: 1},
      {title: 'Declarative Languages', link: '#declarative', subtitle:'subsection', level: 1},
      {title: 'Theory', link: '#theory'},
      {title: 'Broken link', link: '#non-exising section'},
    ]
  },
  {
    title: 'Platforms',
    id: '#platforms',
    links: [
      {title: 'Node,js', link: '#nodejs'},
      {title: 'Frontend Development', link: ''},
      {title: 'iOS', link: ''},
    ]
  },
  {
    title: 'Languages',
    id: '#languages',
    links: [
      {title: 'JavaScript', link: ''},
      {title: 'TypeScript', link: ''},
      {title: 'F#', link: ''},
      {title: 'C#', link: ''},
    ]
  },
  {
    title: 'Theory',
    id: '#theory',
    links: [
      {title: 'Talks', link: ''},
      {title: 'Math', link: ''},
    ]
  },
  {
    title: 'Node,js',
    id: '#nodejs',
    links: [
      {title: 'Node.js topic1', link: ''},
      {title: 'Node.js topic2', link: ''},
    ]
  },
];

export default sections;

export type Section = {
  title: string;
  id: string;
  links: Link[];
}

export type Link = {
  title: string;
  link: string;
  subtitle?: string,
  level?:number;
}