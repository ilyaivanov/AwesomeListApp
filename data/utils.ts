export function validateNonEmpty<T>(value: T | undefined | null, message: string): T {
  if (!value) {
    throw new Error(message);
  }
  else {
    return value;
  }
}

export const isLocalLink = (link: string) => link.indexOf('#') === 0;

export const normalizeLocalLink = (text: string) => '#' + text.toLowerCase().replace(/ /g, '-').replace(/\//g, '');
