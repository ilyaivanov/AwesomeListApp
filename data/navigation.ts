import {Repository, Section} from '../types';
import {normalizeTitle, validateNonEmpty} from './utils';
// import repository from './models/sindresorhus_awesome';
import repository from './models/sindresorhus_awesome-nodejs';

export const navigate = (currentRepository: Repository, link?: string): Section => {
  if (!link) {
    return currentRepository.home;
  }
  // check if link is local
  const section = currentRepository.sections.find(section => normalizeTitle(section.title) === link);
  return validateNonEmpty(section, `Couldn't not found section for link ` + link);
}