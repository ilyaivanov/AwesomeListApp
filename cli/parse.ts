import {Link, Section, Token} from '../data/types';
import * as md from 'markdown-it';
import {find, findIndex, flatten, trim} from 'lodash';
import {normalizeLocalLink, validateNonEmpty} from '../data/utils';
import {allRemoteLinks, createIdForUrl, root, stringify} from './util';
import {mdBase} from './load';
import * as fs from 'fs';
import postProcessing from './postProcessing';

const mark = md({
  html: true,
  linkify: true,
  typographer: true
});

export const parseFromFile = (repoId: string): Section[] => {
  const filePath = mdBase + repoId + '.md';
  console.log(`Reading ${repoId} from ${filePath}`);

  const md = fs.readFileSync(filePath);

  return parseFromMd(md.toString(), repoId);
};

export const parseAll = () => {
  const repSections = parseFromFile(root.id);
  const links = allRemoteLinks(repSections);

  const linkToParse = [0, 1];

  const parsed = flatten(linkToParse.map(i => parseFromFile(links[i])));

  const sections = [
    ...repSections,
    ...parsed,
  ];

  const sectionFormatted = `export default ${stringify(sections)}`;

  fs.writeFileSync('data/sections.ts', sectionFormatted);
  console.log(`Done writing to 'data/sections.ts'`);
};

export const parseFromMd = (md: string, repoId: string): Section[] => {
  const tokens = mark.parse(md, null);
  const headingIndices = tokens
    .map((token, i) => (token.type === 'heading_open' && token.markup !== '#') ? i : -1)
    .filter(i => i >= 0);

  const sections = headingIndices.map(index => localizeSection(parseSection(tokens, index), repoId));

  //first section considered root of the repository, move to postprocessing
  sections[0].id = repoId;
  return postProcessing(sections);
};

const localizeSection = (section: Section, repoId: string) => ({
  ...section,
  id: repoId + normalizeLocalLink(section.title),
  links: section.links.map(l => ({...l, link: localizeLink(l.link, repoId)}))
});


//LEGACY, to refactor
const mapTitle = (token: Token) => {
  const textToken = token.children.find(x => x.type === 'text');
  const validated = validateNonEmpty(textToken, 'Could not find text token in ');
  return validated.content;
};

const parseLink = (token: Token): string => {
  const linkToken = token.children.find(x => x.type === 'link_open');
  if (!linkToken) {
    return '';
  }
  const validated = validateNonEmpty(linkToken, 'Could not find link_open token in ');
  return validated.attrs[0][1];
};

const parseSubtitle = (token: Token) => {
  const textTokenIndex = token.children.findIndex(x => x.type === 'text');
  const subtitleToken = find(token.children, x => x.type === 'text', textTokenIndex + 1);
  const subtitle = subtitleToken ? subtitleToken.content : '';
  return trim(subtitle, '- â€”');
};

const createLink = (token: Token): Link => ({
  title: mapTitle(token),
  subtitle: parseSubtitle(token),
  link: parseLink(token),
  level: Math.min(token.level - 3, 1)
});

const parseSection = (tokens: Token[], startIndex: number): Section => {
  const endParagraphIndex = findIndex(tokens.slice(startIndex), token => token.type === 'heading_close') + startIndex;
  const nextParagraph = findIndex(tokens.slice(endParagraphIndex + 1), token => token.type === 'heading_open') + endParagraphIndex;
  const headerToken = find(tokens, token => token.type === 'inline', startIndex);
  const linkTokens = tokens.slice(endParagraphIndex, nextParagraph).filter(t => t.type === 'inline');
  return {
    id: '',
    title: headerToken ? headerToken.content : 'Unknown',
    links: linkTokens.map(createLink),
  };
};

const localizeLink = (link: string, repoId: string): string =>
  link.startsWith('#') ? repoId + link :  repoId === root.id ? createIdForUrl(link) : link;