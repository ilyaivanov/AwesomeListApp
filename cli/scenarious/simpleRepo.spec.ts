import {Section} from '../../data/types';
import {root} from '../util';
import {parseFromMd} from '../parse';

const rootMd = `
## Contents

- [Platforms](#platforms)
- [Programming Languages](#programming-languages)

## Platforms

- [Node.js](https://github.com/sindresorhus/awesome-nodejs) - JavaScript runtime built on Chrome's V8 JavaScript engine.
- [Frontend Development](https://github.com/dypsilon/frontend-dev-bookmarks)
- [iOS](https://github.com/vsouza/awesome-ios) - Mobile operating system for Apple phones and tablets.
- [Android](https://github.com/JStumpp/awesome-android)

## Programming Languages

- [Node.js](https://github.com/sindresorhus/awesome-nodejs) - JavaScript runtime built on Chrome's V8 JavaScript engine.
- [Frontend Development](https://github.com/dypsilon/frontend-dev-bookmarks)
- [iOS](https://github.com/vsouza/awesome-ios) - Mobile operating system for Apple phones and tablets.
- [Android](https://github.com/JStumpp/awesome-android)
`;

describe('Parsing a sample from Awesome List', () => {
  let sections: Section[];
  beforeEach(() => {
    sections = parseFromMd(rootMd, root.id);
  });

  it('should be matched to a snapshot', function () {
    expect(sections).toMatchSnapshot();
  });

  it('should result in three sections', function () {
    expect(sections).toHaveLength(3)
  });

  describe('Root section', () => {
    let rootSection: Section;
    beforeEach(() => {
      rootSection = sections[0];
    });

    it('should have correct id', function () {
      expect(rootSection.id).toBe(root.id);
    });

    it('should have two link', function () {
      expect(rootSection.links).toHaveLength(2);
    });

  });

  describe('Platforms section', () => {
    let platforms: Section;
    beforeEach(() => {
      platforms = sections[1];
    });

    it('should have the same id as platforms link', function () {
      expect(platforms.id).toBe('sindresorhus_awesome#platforms');
    });

    it('should have four links', function () {
      expect(platforms.links).toHaveLength(4);
    });

    it('should have a link to a node.js repo', function () {
      expect(platforms.links[0].link).toBe('sindresorhus_awesome-nodejs');
    });
  });
});