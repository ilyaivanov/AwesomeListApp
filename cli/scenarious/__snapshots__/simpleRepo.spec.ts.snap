// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing a sample from Awesome List should  match snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should be match snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should be matche snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should be matched  snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should be matched snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should be matched to a snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Contents",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;

exports[`Parsing a sample from Awesome List should match snapshot 1`] = `
Array [
  Object {
    "id": "sindresorhus_awesome",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#platforms",
        "subtitle": "",
        "title": "Platforms",
      },
      Object {
        "level": 0,
        "link": "sindresorhus_awesome#programming-languages",
        "subtitle": "",
        "title": "Programming Languages",
      },
    ],
    "title": "Awesome List",
  },
  Object {
    "id": "sindresorhus_awesome#platforms",
    "links": Array [
      Object {
        "level": 0,
        "link": "sindresorhus_awesome-nodejs",
        "subtitle": "JavaScript runtime built on Chrome’s V8 JavaScript engine.",
        "title": "Node.js",
      },
      Object {
        "level": 0,
        "link": "dypsilon_frontend-dev-bookmarks",
        "subtitle": "",
        "title": "Frontend Development",
      },
      Object {
        "level": 0,
        "link": "vsouza_awesome-ios",
        "subtitle": "Mobile operating system for Apple phones and tablets.",
        "title": "iOS",
      },
      Object {
        "level": 0,
        "link": "JStumpp_awesome-android",
        "subtitle": "",
        "title": "Android",
      },
    ],
    "title": "Platforms",
  },
  Object {
    "id": "sindresorhus_awesome#programming-languages",
    "links": Array [],
    "title": "Programming Languages",
  },
]
`;
